### ISSUE TITLE
Helm upgrades using --wait succeed without readiness checks passing if you downscale replicas to 0

### ISSUE THREAD

**Issue by 94DanielBrown** [helm/helm#12768]
Output of `helm version`:
version.BuildInfo{Version:"v3.13.1", GitCommit:"3547a4b5bf5edb5478ce352e18858d8a552a4110", GitTreeState:"clean", GoVersion:"go1.21.3"}

Output of `kubectl version`:
Client Version: v1.28.3

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
EKS 1.28

I encountered this when testing if kube-downscaler would block pipelines due to scaling down pods before they became ready in downscale enabled environments, but pipelines always succeeded passing the scaled down env where pods never became ready before being terminated in response to kube-downscaler setting the replicas on the deployment to 0. 
In this case helm releases were managed by fluxcd HelmRelease but I've read this should be using helm upgrade --wait.

I recreated by creating a bitnami/nginx helm chart via helm but then adding a non existent readiness check path in the values so that it would never become ready:
readinessProbe:
  enabled: true
  failureThreshold: 3
  httpGet:
    path: /non-existent-path
    port: http
  initialDelaySeconds: 90
  periodSeconds: 5
  successThreshold: 1
  timeoutSeconds: 3
I can make the 'helm upgrade --wait' succeed by simply scaling the nginx deployment to 0 whilst helm is waiting for the readiness checks to succeed.
If I don't scale it down then obviously the helm release fails after the default --wait timeout of 5 minutes.

I haven't looked at the code yet but I could make a PR to make helm fail if readiness checks don't pass, regardless if pods get terminated before the helm timeout elapses by being scaled down by the deployment.
But I'm not sure if this has been considered before and the current behaviour is intended as the preferable way to deal with pods that are scaled down in the midst of a helm upgrade?

**Comment by banjoh** [comment]
Is this a duplicate of https://github.com/helm/helm/issues/12747?

**Comment by z4ce** [comment]
Helm is waiting on the deployment to be ready (like @banjoh points out in the linked issue). What alternate behavior are you proposing?

**Comment by 94DanielBrown** [comment]
> Is this a duplicate of #12747?

It's very similar, changing the code to fix the behaviour here would likely fix that issue too :)

Also I noticed on the other issue there was mention of that example having 1 replica with 1 unavailable. I have tested with multiple replicas and setting maxUnavailable to 0 so this isn't the issue here as when you scale a deployment down to 0 it has status:
Replicas:       0 desired | 0 updated | 0 total | 0 available | 0 unavailable
So because the desired replicas is 0 after scaling it's always 0 unavailable

**Comment by 94DanielBrown** [comment]
> Helm is waiting on the deployment to be ready (like @banjoh points out in the linked issue). What alternate behavior are you proposing?

I can't think of a nice way to make it work with downscaler tbh but I'd prefer if it failed and I had to make a workaround in the downscaler implementation than helm succeeds without pods becoming ready.
The only proposed solution I can think of is perhaps looking at the labels from the deployment and then checking that atleast 1 of or all the initially desired number of pods pods becomes ready.
I don't know how cleanly this can be done and if it's worth it for the downscaler edge case

**Comment by z4ce** [comment]
Maybe the solution would be to disable downscaler during install and then enable it in a post-install hook?

**Comment by 94DanielBrown** [comment]
Yes my solution was going to be to add annotations in the GitHub workflow
to do that when downscaler gets enabled by customer.

I think checking the deployment status is the correct behaviour from helm
and working around the downscaler edge case isn’t a good idea!

I’ll close the issue?

Thanks

**Comment by 94DanielBrown** [comment]
> Maybe the solution would be to disable downscaler during install and then enable it in a post-install hook?

Yes my solution was going to be to add annotations in the GitHub workflow to do that when downscaler gets enabled by customer. 

I think checking the deployment status is the correct behaviour from helm and working around the downscaler edge case isn’t a good idea! 

I’ll close the issue? 

Thanks
